### DISCUSSION TITLE
Coordinate ordering?

### DISCUSSION THREAD

**Discussion by jaideep11061982** [rwightman/efficientdet-pytorch#232] [Q&A]
**Describe the bug**
I tweak the order of  box coordinates in dataset for target from xyxy to yxyx 
```
sample1=sample.copy()
print(boxes)#yxyx
for box in boxes:
    cv2.rectangle(sample,
                  (box[1], box[0]),
                  (box[3], box[2]),
                  (220, 0, 0), 2)
    
#ax.set_axis_off()
ax.imshow(sample)
```
output
```
[[122  63 371 120]
 [162 365 392 444]]
```
When i check the models predictions for corresponding image it seem to be giving in the order xyxy
```
for box in result_boxes[1]:
    print(box)
    cv2.rectangle(sample1,
                      (int(box[0]), int(box[1])),
                      (int(box[2]), int(box[3])),
                      (220, 0, 0), 2)
plt.imshow(sample1,figure=(12,8))
```
```
[ 74.547134 116.86316  147.59622  340.9181  ]
[355.19867 136.82175 426.5904  372.49075]
```
I get this. Visually  predicted and target look identical  .

**To Reproduce**
1.
2.

**Expected behavior**
Why the order of model is different than what we pass in target. 

Also fyi . My mAP even though for a sample image prediction look good   never cross up 0.02 . 
m i getting confused with box orders ?

**Screenshots**
If applicable, add screenshots to help explain your problem. predicted
![image](https://user-images.githubusercontent.com/38164196/122050517-266e8b00-ce01-11eb-8377-64269d4a394c.png)

Ground truth
![image](https://user-images.githubusercontent.com/38164196/122050593-3a19f180-ce01-11eb-9b1a-5416630860ae.png)

**Desktop (please complete the following information):**
 - OS: [e.g. Windows, Ubuntu 18.04]
 - This repository version [e.g. pypi 0.3.1 or branch/commit ref]
 - PyTorch version w/ CUDA/cuDNN  [e.g. from `conda list`, 1.7.0  py3.8_cuda11.0.221_cudnn8.0.3_0]

**Additional context**

**Comment by rwightman** [comment]
@jaideep11061982  The processed output from the model (after NMS) is XYXY, torchvision's NMS operates in XYXY so I convert from the model's YXYX and I decided not to convert back (for performance, since COCO needs it in XYWH anyways). It's mentioned in a few places and you can search past issues, discussions. But not all that clear I guess.

The internal model and anchor / labelling code based on the original TF model layout (and for weight compatibility) is all YXYX so that is why the targets need to be YXYX. The evaluators also differ, COCO evaluator expects XYWH and the TF models based evaluators tha are used for Pascal and OpenImages expect YXYX. Conversions are all done correctly in the example COCO / Pascal / OpenImages datasets, train/validation scripts, and Evaluators.

**Reply by jaideep11061982** [reply]
@rwightman  Thanks
I kind of get this after digging into the code. 
1)  confirming what you said
say
```
net= model_from_config()
net(image,target) ->  xyxy    (post processing results into xyxy) 
net.model(image) - >yxyx  -> post processing in detbranch-> xyxy ?
```
Please let me know if I got you correct 

2) If i have to compute the IOU/maP  between GT and Predictions , do i have  to convert GT  back to xyxy  for pascal voc format?

3) Box Loss computation would be still using yxyx format 
` loss, class_loss, box_loss = self.loss_fn(class_out, box_out, cls_targets, box_targets, num_positives)`

4) on a other note  about class setting , Do i have to set No finding class inside dataset to -1 ?

5) Lastly , which is confusing . 
Num_class passed to model , if suppose i have 1 no finding class  and 3 finding class
then  the num classes passed to model is  4 or 3 

Last i used effdet for a problem which had no No finding class, this time  both are there, model results are below par  so that is why so many question.. Please bear with me :)

**Reply by rwightman** [reply]
@jaideep11061982 many of the defaults were adopted from the original TF impl...

The 'background' class (no finding as you call it), is 0 from the perspective of your dataset, so the first valid class label your dataset should produce is '1', generally you don't need to explicitly provide background labels but I suppose you could in some cases. You can see the demo datasets add the 1 offset when returning the target labels.

Internally the anchor labelling code subtracts that offset so the background ends up being -1 and -2 is used as a sentinel for some other things, the non-processed output of the model does not include the background (it's used in the anchor labelling/matching) so the `num_classes` you build the model with should NOT include the background class, ie in the Pascal VOC demo in README you pass in 20 classes, not 20 + 1.

And yes, you are correct, the bench(input, target) outputs xyxy, the bench.model -> yxyx, and there is a step in the post processing that the bench does right before NMS where it's switched to xyxy and left that way. Losses are not calculated on the post-process output so they all remain in yxyx.

What format you need for your evaluation depends on the evaluation code you use, as I mentioned the included generic TF based evaluators that are used for the built in VOC support expect yxyx internally (https://github.com/rwightman/efficientdet-pytorch/blob/master/effdet/evaluator.py#L121-L175), the `pred_yxyx=` arg specifies whether you will pass yxyx coords to the evalutor (instead of xyxy) and it adjusts.
